File: java/VendingMachine/test/com/ravendarque/unitTests/RailsShould.java
package com.ravendarque.unitTests;

import com.ravendarque.vendingMachine.rails.*;
import org.junit.jupiter.api.Test;

import java.util.Map;

import static org.junit.jupiter.api.Assertions.*;

class RailsShould {

    private static final String TEST_RAIL_CODE = "T1";

    @Test
    void returnFalseWhenCheckingIfNonexistentRailCanBeSelected() {

        final String testNonExistentRailCode = "XX";

        assertFalse(getTestRails().canSelectRail(testNonExistentRailCode));
    }

    @Test
    void returnTrueWhenCheckingIfExistentRailCanBeSelected() {

        assertTrue(getTestRails().canSelectRail(TEST_RAIL_CODE));
    }

    @Test
    void selectRailByRailCode() {

        final String expectedLabel = "Test label";

        final Rails testRails = getTestRails();
        testRails.selectRail(TEST_RAIL_CODE);

        final Rail actualSelectedRail = testRails.getSelectedRail();

        assertEquals(expectedLabel, actualSelectedRail.getLabel());
    }

    @Test
    void returnRailsSummary() {

        final String expectedRailLabel = "Test label";

        Rails testRails = getTestRails();
        Map<String, String> actualRailsSummary = testRails.getRailsSummary();

        assertEquals(actualRailsSummary.get(TEST_RAIL_CODE), expectedRailLabel);
    }

    private Rails getTestRails() {

        final int testPrice = 1;
        final int dummyInitialInventory = 1;
        final String testLabel = "Test label";

        final RailsConfiguration railConfiguration = new RailsConfigurationBuilder()
                .add(new RailConfigurationSettings(
                        TEST_RAIL_CODE,
                        testPrice,
                        dummyInitialInventory,
                        testLabel))
                .build();

        return new Rails(railConfiguration);
    }
}

File: C++/VendingMachine/VendingMachine.UnitTestsG/RailsShould.cpp
#include "pch.h"
#include "../VendingMachine/Rails.h"
#include "../VendingMachine/RailsConfiguration.h"

TEST(RailsShould, ReturnFalseWhenCheckingIfNonexistentRailCanBeSelected)
{
	const std::string testNonExistentRailCode = "XX";

	const RailsConfiguration testRailsConfiguration;

	const Rails rails(testRailsConfiguration);
	const auto actualResult = rails.CanSelectRail(testNonExistentRailCode);

	EXPECT_FALSE(actualResult);
}

TEST(RailsShould, ReturnTrueWhenCheckingIfExistentRailCanBeSelected)
{
	const std::string testRailCode = "T1";
	const unsigned int dummyCapacity = 0;
	const double dummyPrice = 0;
	const unsigned int dummyInitialInventory = 0;
	const std::string dummyLabel = "Dummy label";
	const RailConfigurationSettings testRailConfigurationSettings(testRailCode, dummyPrice,
	                                                              dummyInitialInventory, dummyLabel);

	RailsConfiguration testRailsConfiguration;
	testRailsConfiguration.Add(testRailConfigurationSettings);

	const Rails rails(testRailsConfiguration);

	const auto actualResult = rails.CanSelectRail(testRailCode);

	EXPECT_TRUE(actualResult);
}

TEST(RailsShould, SelectRailByRailCode)
{
	const std::string testRailCode = "T1";
	const unsigned int dummyCapacity = 0;
	const double dummyPrice = 0;
	const unsigned int dummyInitialInventory = 0;
	const std::string testLabel = "Test label";
	const RailConfigurationSettings testRailConfigurationSettings(testRailCode, dummyPrice,
	                                                              dummyInitialInventory, testLabel);

	RailsConfiguration testRailsConfiguration;
	testRailsConfiguration.Add(testRailConfigurationSettings);

	Rails testRails(testRailsConfiguration);
	testRails.SelectRail(testRailCode);

	const auto actualSelectedRail = testRails.GetSelectedRail();

	EXPECT_EQ(actualSelectedRail->GetLabel(), testLabel);
}

TEST(RailsShould, ReturnRailsSummary)
{
	const std::string testRailCode = "T1";
	const unsigned int dummyCapacity = 0;
	const double dummyPrice = 0;
	const unsigned int dummyInitialInventory = 0;
	const std::string testLabel = "Test label";
	const RailConfigurationSettings testRailConfigurationSettings(testRailCode, dummyPrice,
	                                                              dummyInitialInventory, testLabel);

	RailsConfiguration testRailsConfiguration;
	testRailsConfiguration.Add(testRailConfigurationSettings);

	const Rails testRails(testRailsConfiguration);

	auto actualRailsSummary = testRails.GetRailsSummary();

	for (const auto& railSummary : actualRailsSummary)
	{
		EXPECT_EQ(railSummary.first, testRailCode);
		EXPECT_EQ(railSummary.second, testLabel);
	}
}

TEST(RailsShould, ReturnFalseWhenCheckingIfARailHasBeenSelectedButNoneHas)
{
	const RailsConfiguration dummyRailsConfiguration;
	const Rails testRails(dummyRailsConfiguration);

	ASSERT_FALSE(testRails.HasSelectedRail());
}
